### DISCUSSION TITLE
Invalid implicit conversion from long to int

### DISCUSSION THREAD

**Discussion by myblindy** [dotnet/csharplang#8201] [General] [answered]
**Version Used**:  C# 12, though it's present in all of them

**Steps to Reproduce**:

```
public struct UnixTime
{
    public static explicit operator UnixTime(int i) => default;
}

public class C {
    public UnixTime M(long i) =>
        (UnixTime)i;
}
```

[Demo here](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQMwAIUIBOArgMYK4CqMUAHgCpRgCmAsAFADeHuvuO/AIwA2XM1oAHADZRSUCgHsJzQhAQLCVGgybMAFLApQAlLgC8APlwATZgDMIxKQgDcHAL4cOAzACZcAMK43Ox8/HjUdIwsuACyelIKMADmuCbmFjxhfHqROizGUG7s7kA===)

**Diagnostic Id**:

None

**Expected Behavior**:

As per [the built-in numeric conversion rules](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/numeric-conversions#implicit-numeric-conversions), the code above should fail.

**Actual Behavior**:

Instead it inserts an implicit conversion from `long` to `int`, then matches the explicit operator's signature:

```
        IL_0000: ldarg.1
        IL_0001: conv.i4
        IL_0002: call valuetype UnixTime UnixTime::op_Explicit(int32)
        IL_0007: ret
```

**Comment by gurustron** [comment]
As far as I understand this is not a bug but behavior defined in the spec ([10.5.5 User-defined explicit conversions](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/conversions#1055-user-defined-explicit-conversions)) which allows using intermediate  standard explicit conversions in this case.

**Comment by RioMcBoo** [comment]


**Reply by jilles-sg** [reply]
The confusing behaviour follows the specification at https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/conversions#1055-user-defined-explicit-conversions . The idea appears to be that a standard explicit conversion can be inserted both before and after a user-defined explicit conversion operator, but the types should not be "too unrelated" due to the part "Find the set of applicable user-defined and lifted conversion operators". So casting a `long` to `UnixTime` works because there is an implicit conversion from `int` to `long` and an explicit conversion from `long` to `int`, but casting an `ulong` to `UnixTime` does not work because there are no implicit conversions between `int` and `ulong` in either direction.

I suggest defining the conversion from `int` to `UnixTime` as an extension method, which provides both the requested type-safety and readable left-to-right operation. The conversion from `UnixTime` to `DateTime` can be an instance method.

By the way, 2038 is less than 13 years away, so `long` would be a better underlying type for `UnixTime`.

**Reply by RioMcBoo** [reply]
I deleted my old misleading messages and added new correct descriptions (please look this thread again)

**Comment by RioMcBoo** [comment]
I think many people don't understand what I'm talking about. The first message in this discussion gave an example that doesn't directly address the current problem.

And I think no one read my messages carefully after that and jumped to conclusions automatically, not realizing the problem.

Now I'll redo this example, and you'll finally understand what I was trying to convey.

A brief background: I asked a question on the StackOverFlow forum, and a girl confirmed that she thought there was a problem. And I really thought that this was the problem. She created this thread and gave me a link to it. So, all the discussion remained behind the scenes, except for her message, which doesn't address the essence of the problem. Simply because if the C# analyzer behaved consistently and identically for all data types, I would have guessed myself that these are the rules of explicit type conversion.
But C# behaves differently depending on the types.
Look at the following example. 


```
public struct UnixTime
{
    public static explicit operator UnixTime(int i) => default;
}

public class C
{
    public UnixTime Method(ulong i) =>
        (UnixTime)i;    // type narrowing (ulong -> int) not allowed

    public UnixTime Method(long i) =>
        (UnixTime)i;    // type narrowing (long -> int) allowed
}

```
<img width="592" height="275" alt="image" src="https://github.com/user-attachments/assets/e19a23fe-9e33-4e09-b571-230238408f58" />

<img width="618" height="317" alt="image" src="https://github.com/user-attachments/assets/dc9c3205-f2e4-4a12-ae57-8e5cbe8855ac" />


Now you can clearly see that the example in the first message does indeed work according to the language rules. I used an explicit conversion, knowing it was a narrowing type, which meant I was accepting the loss of data (long -> int).

But from another perspective, if I perform the exact same explicit conversion and also agree to a type narrowing (ulong -> int), C# assumes that even though I used an explicit conversion, I must adhere to the type specification. (And this, in my opinion, is correct, because that's what strong typing is for, and explicit conversion for base types works as in the first case because the operator actually calls overloaded functions for the specific base type and the target base type.)

**Reply by RioMcBoo** [reply]
Then I started asking why this was happening, because if you just take the basic types and try to do an explicit conversion, there won't be any errors:

```
public class C
{
    public int Method(ulong i) =>
        (int)i;    // type narrowing (ulong -> int) allowed

    public int Method(long i) =>
        (int)i;    // type narrowing (long -> int) also allowed
}
```

<img width="536" height="153" alt="image" src="https://github.com/user-attachments/assets/53d8aa07-b7e9-4046-80a5-20548b469755" />


[I indirectly concluded that this occurs because all basic integer types share a common type to which they are implicitly converted—float/double](https://github.com/RioMcBoo/Why-.NET-has-wrong-implicit-conversations). The error supposedly occurs because float/double support negative numbers, just like int/long.

But when it comes to unsigned numbers, the implicit type conversion to float/double doesn't work, which is why the error occurs.

To prove this, I tried converting an int to a uint . I expected both explicit conversions to fail. But to my surprise, the opposite happened—both explicit conversions succeeded.

```
public struct UnixTime
{
    public static explicit operator UnixTime(uint i) => default;
}

public class C
{
    public UnixTime Method(ulong i) =>
        (UnixTime)i;    // type narrowing (ulong -> uint) allowed

    public UnixTime Method(long i) =>
        (UnixTime)i;    // type narrowing (long -> uint) also allowed
}
```

<img width="595" height="293" alt="image" src="https://github.com/user-attachments/assets/f6a07ebc-2d7b-4638-8da0-2d96439c0288" />


In any case, this means that the behavior of explicit type conversions is based on some kind of inherent logical error. And I can only guess that it's related to the implicit conversion to float/double.

However, I believe that user-overridden conversion operators should throw an error even when explicitly called, because users might make a mistake when defining their own types and not notice that data has been lost.

For base types, this error can be omitted, as everyone learning the language knows this rule. However, no one knows the behavior of your class or structure until they look at the code, so it's a non-obvious operation, and the user must strictly adhere to all types. This is my humble opinion.

So, to summarize:
1) There is a bug and it needs to be fixed.
2) We need to define explicit and implicit conversion rules for overloaded conversion operators, and preferably move toward stricter type checking.

**Reply by jnm2** [reply]
Your guess that it's related to implicit conversion to float/double is not correct. Walking through the spec (https://github.com/dotnet/csharplang/discussions/8201#discussioncomment-15118618), it is evident that the conversion from `long` works only because there is a standard implicit conversion from `int` to `long`, and likewise the conversion from `ulong` does not work because there is no standard implicit conversion from `int` to `ulong`.

I haven't yet been able to form an intuition for why it was done on the basis of a _reverse_ standard implicit conversion, rather than, say, allowing the explicit user-defined conversion on the basis of a a standard _explicit_ conversion from `long` to `int` and from `ulong` to `int`, which do both exist.

(Apologies for my earlier message which had a different conclusion; I majorly misread the spec at first!)

Given that this thread was created as a bug report about `long` and not about `ulong`, regardless of how and why this thread was started, I would recommend pursuing any language changes in a brand new thread.

**Reply by RioMcBoo** [reply]
That's the problem, in my opinion. I described the issue broadly. Reporting bugs separately for each type and each case won't be obvious, nor will it be clear how these issues are interconnected.

It's better to simply rename this discussion to something else that can more broadly describe what's going wrong.

Besides, I don't like the very approach of trying to conform to some specification.

This way, you can describe all the bugs in the specification and say they're not bugs, but intended that way.

It's very disappointing that this issue is being held back due to the need to maintain already compiled code.

I'd like to see a new version of the compiler in the future that will fix known bugs, with the ability to select the version in the settings for new projects, so as not to depend on legacy code.

**Reply by RioMcBoo** [reply]
I mean that the specification corresponds to reality, not that reality should correspond to the specification.
For example, I don't understand why C# allows conversion from long to double/float/etc., losing data in the process. Even if it's described in the specification, we need to ask why it was done in the first place if it doesn't align with common sense. It was probably originally done by mistake, and then added to the specification to legitimize this behavior.

**Reply by jnm2** [reply]
My thought was to give you the best chance of being heard out, since you're coming with a request for a language change, whereas this thread thus far has not been about requesting a language change. It was a clear report that the compiler didn't match the spec, with the clear resolution that it actually does match the spec. Starting with a clean slate will prevent every reader from starting at the top and misunderstanding what the whole thread is about, and I don't really see the downside to giving yourself that clarity. However if you would rather ignore this and take your chances at redefining the thread, I hope it works for you!

We can do both things at once: both keep a compiler conforming to a specification, and evolve that specification forward. In practice, these goals don't end up in conflict. It's good to be clear on what the specification actually says before proposing changes to it. I don't see a way in which your issue is being held back; in fact, my goal was to get your issue moving by suggesting starting a new thread.

**Reply by RioMcBoo** [reply]
If you know how to do this (in which project branch, with what tags, and how to name it correctly), I'd ask you to take on this burden, as my understanding of the compiler's internals is limited, and my English is also poor. All my attempts to provide feedback are usually misunderstood, or I'm doing it incorrectly.

Even now. It looks like some guy off the street has decided to change the way C# works globally. It seems very audacious, but deep down, I know it shouldn't work this way.

I merely pointed out the location, the symptoms of the problem. Perhaps the problem is much deeper, and if specialists look into it, they'll fix it globally, not just for the long type.

I deleted the post (you can perhaps view the history) where I listed all the types that experience the same issue. It's wrong, in my opinion, to only address the reported issues and ignore the underlying problem.

So, if I create a new topic and describe the entire list of types this happens to, but I forget one type through mistake or carelessness, will you correct all the types except the one I forgot?

I can't understand how this machine called "corporation" works.

**Reply by jnm2** [reply]
I would recommend starting a thread in the Discussion section of this repository, with the category Language Ideas, with a title that describes the underlying problem itself, as you see it, which contains a few concrete examples, just enough to illustrate the range of the underlying problem, and which speaks to the concrete upsides of fixing the problem.

Be ready for community members to weigh in who want to explore alternatives or to confirm why it's actually worth doing. This is a normal part of our community here. If the idea is really a good idea for the whole community, the path to getting traction is to engage with all the exploration- the questions, concerns, and brainstorming, with a "yes, and" mentality. You don't have to be the one to carry it forward, you can respond or not with as much or as little time or energy as you want to put in.

Yeah, corporations can be a weird phenomenon, right? If it helps to know, I'm a community member outside of Microsoft and I'm volunteering my spare time to help move your ideas forward.

**Comment by jnm2** [comment]


**Reply by 333fred** [reply]
Note that this is a native compiler compat behavior; you can verify this by trying out the native compiler yourself on a Windows machine by compiling the code with `C:\Windows\Microsoft.NET\Framework64\v4.0.30319\csc.exe`. We cannot fix this, as it will break too much real code.

**Comment by jnm2** [comment, marked as answer]
This is not a compiler bug, but is how explicit conversions have been specified since C# version 1.

I'm a little confused about why the spec checks whether a standard implicit conversion (`int` to `long`) exists in _reverse_ during an explicit conversion, rather than checking whether standard _explicit_ conversions exist such as `ulong` to `int`. I'm curious if anyone has an intuition for what the spec is doing here.

(The spec does eventually, by implication, require the explicit conversion to _also_ exist which matches the reverse implicit conversion, because the explicit conversion is what is performed if the matching reverse implicit conversion exists.)

## Walkthrough

This is the reported sample:
```cs
public struct UnixTime
{
    public static explicit operator UnixTime(int i) => default;
}

public class C
{
    public UnixTime M(long i) => (UnixTime)i; // This cast should fail to compile, but it compiles successfully.
}
```

`(UnixTime)i` is a user-defined explicit conversion. [§10.5.5 User-defined explicit conversions]( https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/conversions.md#1055-user-defined-explicit-conversions) defines how this is processed.

> A user-defined explicit conversion from an expression `E` to a type `T` is processed as follows:

`E` is `i` and `T` is `UnixTime`.

> - Determine the types `S`, `S₀` and `T₀`.
>   - If `E` has a type, let `S` be that type.

`i` has the type `long`, so `S` is `long`.

>   - If `S` or `T` are nullable value types, let `Sᵢ` and `Tᵢ` be their underlying types, otherwise let `Sᵢ` and `Tᵢ` be `S` and `T`, respectively.
>   - If `Sᵢ` or `Tᵢ` are type parameters, let `S₀` and `T₀` be their effective base classes, otherwise let `S₀` and `T₀` be `Sᵢ` and `Tᵢ`, respectively.

These special cases don't apply; `S₀` remains `long` and `T₀` remains `UnixTime`.

> - Find the set of types, `D`, from which user-defined conversion operators will be considered. This set consists of `S₀` (if `S₀` exists and is a class or struct), the base classes of `S₀` (if `S₀` exists and is a class), `T₀` (if `T₀` is a class or struct), and the base classes of `T₀` (if `T₀` is a class). A type is added to the set `D` only if an identity conversion to another type already included in the set doesn’t exist.

The set of types `D` thus consists of `long` and `UnixTime`.

> - Find the set of applicable user-defined and lifted conversion operators, `U`. This set consists of the user-defined and lifted implicit or explicit conversion operators declared by the classes or structs in `D` that convert from a type encompassing `E` or encompassed by `S` (if it exists) to a type encompassing or encompassed by `T`. If `U` is empty, the conversion is undefined and a compile-time error occurs.

According to this step, the compiler must include in `U` the user-defined conversion operator `UnixTime(int i)`, because it converts from a type (`int`) which is encompassed by `S` (type `long`).

"Encompassing" is defined in [§10.5.3 Evaluation of user-defined conversions](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/conversions.md#1053-evaluation-of-user-defined-conversions):

> - If a standard implicit conversion ([§10.4.2](conversions.md#1042-standard-implicit-conversions)) exists from an expression `E` to a type `B`, and if neither `B` nor the type of `E` (if it has one) are *interface_type*s, then `E` is said to be *encompassed by* `B`, and `B` is said to *encompass* `E`.

Because there is a standard implicit conversion from the type `int` to the  type `long`, the type `long` encompasses the type `int`. The user-defined conversion operator in the sample converts from `int`, which is a type that is encompassed by `S`, so the user-defined conversion operator is included in set `U`, rather than being excluded as the original post expected.

(This is the point at which the algorithm would diverge for `ulong`. `ulong` does not encompass `int` because there is no implicit conversion from `int` to `ulong`. This is because `ulong` cannot represent negative `int` values. For an `S` of `ulong` rather than `long`, the algorithm would terminate here with an empty `U` causing a compile-time error.)

**Comment by Nigel-Ecma** [comment]
@myblindy & @RioMcBoo – Can either of you please post the link to the original StackOverflow question (my Google-foo is failing me)?

What you have reported _**is**_ a language/compiler/Standard bug and I want to try to get it fixed.

**Reply by RioMcBoo** [reply]
I could not find it :(
Anyway, it will be helpless. 

My last message clearly shows what is the problem. 

It should have initiated an error in any way (following human logic about what the program language is and how it must help people to avoid mistakes) if we don't have overloaded operator ulong to int in current showed case as like long to int either. 

But people from Microsoft said that it is impossible because it will ruin a lot of "no errors after compiling code" that runs already.

So there are only two ways to do it:
1) make ruining fix
2) allow any explicit double conversion  for any basic type to save any logic in conversion rules.

The last one could also initiate unexpected behaviour and errors related to undefined dual behaviour due to the fact that two valid paths have appeared from which the compiler itself cannot choose the correct method overload. 

So this is deadlocked situation that only could be managed by Microsoft, but they have easiest way - find a new loophole in specification to say "this is not a bug but a feature"

**Reply by Nigel-Ecma** [reply]
@RioMcBoo & @myblindy – Thanks for looking RioMcBoo. Don’t be too negative yet, but no promises.

**Reply by RioMcBoo** [reply]
2) allow any explicit double conversion  for any basic type to save any logic in conversion rules.
- i meant literally doubled conversion , not related to "double" type of data.

As you can see - we even cannot make a point is that is good in this case or not. That because i related to human logic of it. 

In stack overflow i reported about my current technician problem with my custom time-data conversions , i could not explain what happened there and asked about it.

So she found in IL code the fact about narrowing long to int and created this theme.

All this time i tried to imagine the root problem of it and find some human logic to legit it. But it is really looks like a bug or something. 


And i really sory about my aggressive manner of speaking, this is because my native language is Russian. The entire style of this language based on emotional expression. And when i try to be neutral - i find my own thoughts were not be translated as i wanted or i must broke my mind. 

Maybe it is just lack of softskills, i cannot fix myself

**Reply by AartBluestoke** [reply]
@RioMcBoo - the following code compiles because of the allowed ulong to uint narrowing - an extra cast (of course one that did work not just going default) could allow ulong to work, as per the examples above .. but unless you have long, uint, and int - you'll get one of ulong and long to break  

```
public struct UnixTime
{
//    public static explicit operator UnixTime(long i) => default;
    public static explicit operator UnixTime(int i) => default;
    public static explicit operator UnixTime(uint i) => default;
}

public class C
{
    public UnixTime M(ulong i) => (UnixTime)i; // via uint
    public UnixTime M(long i) => (UnixTime)i; // This cast now fails as ambiguous
}
```

    uint to ulong is an [allowed converstion](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/conversions.md#1023-implicit-numeric-conversions)

i agree that the surprise narrowing with no warning is surprising ...

**Reply by RioMcBoo** [reply]
> @RioMcBoo - would option 3, working, be to add a uint option here - you can't express time in the past, but if you know you are talking about "now-ish" times, then that would be fine:
> 
> ```
> public static explicit operator UnixTime(uint i) => default; 
> uint to ulong is an [allowed converstion](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/conversions.md#1023-implicit-numeric-conversions)
> ```

Could you provide more info? i didn't catch what past\future time you wrote about.

And i didn't find any possible cases when hardcoding explicit operator to return "0" could help to avoid some errors instead of make them.

Anyway, my point of interest is making clear rule of conversions and avoiding mistakes, not finding some solution to manage current problem with my custom time-data types. I could manage with that issue, i just wanted to make a signal about something in C# goes wrong .

**Reply by RioMcBoo** [reply]
> @RioMcBoo - the following code compiles because of the allowed ulong to uint narrowing - an extra cast (of course one that did work not just going default) could allow ulong to work, as per the examples above .. but unless you have long, uint, and int - you'll get one of ulong and long to break
> 
> ```
> public struct UnixTime
> {
> //    public static explicit operator UnixTime(long i) => default;
>     public static explicit operator UnixTime(int i) => default;
>     public static explicit operator UnixTime(uint i) => default;
> }
> 
> public class C
> {
>     public UnixTime M(ulong i) => (UnixTime)i; // via uint
>     public UnixTime M(long i) => (UnixTime)i; // This cast now fails as ambiguous
> }
> ```
> 
> ```
> uint to ulong is an [allowed converstion](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/conversions.md#1023-implicit-numeric-conversions)
> ```
> 
> i agree that the surprise narrowing with no warning is surprising ...
```
public struct UnixTime
{
    private ticks;
    public UnixTime(int i) { ticks = i; }

    public static implicit operator int(long i) => (int)i; // 1st implicit conversion

    public static explicit operator UnixTime(int i) => new UnixTime(i); // 2nd implicit converion (explicit counts as implicit in that case)


}

public class C
{
    public UnixTime Method(long i) =>
        (UnixTime)i;
}

```
When i wrote about doubled conversion I meant these:
1) long->int
2) int->UnixTime

If both of implicit conversions is implemented - it should work. In other cases - trow error. 

How it works now:
1) sometimes goes automatically without any implementation and avoid base implicit conversions rules. 
2) goes well

Very these two steps i called "double conversion". 

And first time i thought it was related to implicit int->single \ long-> single conversions. But i was wrong, especially because it would be triple conversion when compiler has ambiguous behaviour. But also i provided couple cases when this logic was wrong. See above.

Notes:
I wrote from my phone and cant check my examples. I can be wrong about triple implicit conversion cause to ambiguous errors. Maybe it works with double implicit conversion. In that case we must have only one implemented implicit conversion and one explicit one to allow compiling. Anyway explicit conversion counts as implicit so i changed my post due to this note.

**Reply by jnm2** [reply]
> Maybe it is just lack of softskills, i cannot fix myself

It is heartwarming to hear someone share such poignant self-reflection! If I may, I would also offer encouragement. You're exactly right that you can't fix yourself, and the good news is you don't need to fix yourself! The lack of softskills is not "you," and so there is no "self" to worry about fixing. Through merely being aware, the transformation is already in progress. Happy journeying!

**Reply by Nigel-Ecma** [reply]
@myblindy & @RioMcBoo – Took longer than planned but the issue is now submitted to [csharpstandard](https://github.com/dotnet/csharpstandard/issues/1528) and [csharplang](https://github.com/dotnet/csharplang/discussions/9917) (full issue in the former, linking issue in the latter). I have reported it as a design bug that needs fixing, we’ll see if it is fixed or the Ford Pinto strategy is adopted…

*Disclaimer*: You are credited with reporting the issue, while I take full blame for the issue submissions to csharpstandard & csharplang – if any bouquets arrive as well as brickbats the former are yours!
